### DISCUSSION TITLE
Use different Classic scoring formula per ruleset

### DISCUSSION THREAD

**Discussion by respektive** [ppy/osu#17125] [Ideas] [answered]
While the current Classic Scoring formula works really well for osu! (and to some extend catch aswell), is it very inflated for taiko and obviously mania.  
It would make sense to use different formulas per ruleset to get them closer to stable. I'm interested to hear some opinions on this and wether this is wanted or if the direction is to use the same Classic Scoring for every ruleset.

**Comment by WitherFlower** [comment]
This is related to this comment : https://github.com/ppy/osu/issues/14794#issuecomment-1059912379 explaining that Classic scoring is way off for taiko because it is linear in stable instead of quadratic

**Comment by bdach** [comment]
one of the goals with scoring for lazer was to make it simpler across the board. adding back ruleset-specific exemptions goes against that explicitly.

**Reply by peppy** [reply]
If the score doesn't grow fast enough we could introduce a single per ruleset multiplier similar to what is done with the ratios in standardised.

**Reply by respektive** [reply]
We already have a multiplier being applied to the score to bring it more towards scorev1 values, that could potentially be altered per ruleset, instead of being a fixed value.

**Reply by Epicm14** [reply]
Hi I did some testing and don't believe that a single algorithm with multipliers really works that well since each mode is scaled differently with combo and density. Especially mania and taiko shouldn't be calculated as combo, even if the goal is simplicity.
I found that it was necessary to tweak some values per mode, such as decreasing the factor and increasing the multiplier for more accuracy based modes.
As it is right now with the current ^2 factor, every mode will gain quadruple score for every time hit count doubles. With this, the current ruleset multiplier for mania suggested my respektive gives mania's longest map of 38,000 hit objects 18 billion score. So I thought the formulas for mania and taiko should be closer to ^1 or linear. Also added a factor to the standardized score ratio to make mania and taiko score scale better with accuracy(near perfect scores give more while mistakes are more unforgiving).
Algorithms I'm proposing:
![image](https://user-images.githubusercontent.com/38800690/158072936-1c4d1893-a69c-4d0d-993e-8dfdcf222955.png)
I thought star rating should be included in the formula as well but if that's too much that's fine. 
![image](https://user-images.githubusercontent.com/38800690/158073336-a9673038-ca2f-4c9b-8200-359bd359a9a4.png)
These values look very satisfying to me and I hope you consider having a separate algorithm per mode(or just mania at the very least), unless you would want to adopt a more linear approach for every mode such as this, but then that would mess with std and ctb.
I also talked to people who know a bit more about this stuff and they agree that trying to balance each mode with a single algorithm isn't really a good idea unless something fundamental about scoring changes.

**Reply by bdach** [reply]
classic scoring is supposed to _match stable first and foremost_. it is not supposed to be "good". the "good" scoring is standardised.

and at this point where all of this complexity is being reintroduced i'd rather just reimplement stable scoring 1:1 and call it a day.

**Reply by smoogipoo** [reply]
> I thought star rating should be included in the formula as well but if that's too much that's fine.

Star rating can't be included. At least not PP-SR. osu-stable uses some different ["difficulty multiplier"](https://osu.ppy.sh/wiki/en/Gameplay/Score/ScoreV1#how-to-calculate-the-difficulty-multiplier) for ScoreV1 based on beatmap attributes including CS, OD, drain length and number of objects, but for the sake of simplicity I think even this is too much to include for the time being because it becomes difficult to figure out these values without needing a significant amount more data. I don't think, for example, that the API provides the drain length, nor the post-mod CS/OD values, etc...

@Epicm14: How do you propose your algorithms to look without the star rating multiplier?

I don't think this is such an outlandish idea, but I'm not entirely sure what this is achieving. As far as I understand @respektive's changes were about balancing inter-ruleset scoring, so that maps in osu! that give 50M score don't give 1T score in osu!mania, for example.
Is this aimed at doing the same thing / around the same line of thinking, or is it something else?

**Reply by Epicm14** [reply]
Just like this:
![image](https://user-images.githubusercontent.com/38800690/158101457-d0c85706-224e-422c-bd93-5df68bccdc4d.png)
Standard and Ctb can be left the same because they work so well already.
It's pretty much aimed at doing the same thing. I discussed this with respektive a bit and we agreed attempting to balance doesn't really work as well as previously thought. This aims to give mania and taiko a more reasonable way to balance scoring across modes. Similar to how taiko's scoring currently is in stable.

**Reply by peppy** [reply]
If possible, could you please link the spreadsheet or provide the formulas as a code block so we can work with them?

```md
osu!: 36*(Standardized Score/1,000,000)* Hitcount)^2
osu!mania: 3000*((Standardized Score/1.000.000) ^2*Hitcount)^1.3
osu!taiko: 3000*((Standardized Score/1,000.000)^1.3*Hitcount)^1.3
osu!catch: 28*(Standardized Score/1,000,000)*Hitcount)^2
```

**Reply by Epicm14** [reply]
I'm not too familiar with github so here you go:
https://docs.google.com/spreadsheets/d/1yM65zGwnv0sVk9voPDha_NGbbxabTHFGo08oho6W0rE/edit?usp=sharing
Thanks for consideration.
Like what respektive did with the multipliers, they could be tweaked here as well for further improvements if necessary.

**Reply by respektive** [reply]
> As far as I understand @respektive's changes were about balancing inter-ruleset scoring, so that maps in osu! that give 50M score don't give 1T score in osu!mania, for example.

That's right. Though i noticed later on that at least for mania it doesn't work that well, because mania on average has so many more objects and is also the only ruleset (for now anyway) that can have more than 1 object at any point in time. So an average 5min marathon in osu might have around 1500 objects, but in mania it can be up to like 5000.

This was kinda overlooked in my testing by mainly checking converts and shorter mania  maps. 

After talking to @Epicm14 and having a [small poll on twitter](https://twitter.com/respektivee/status/1501780553925804033), we came to the conclusion, that increasing the score values from scorev1 for mania/taiko is appreciated, but not as much as i was proposing by bringing them up to osu/catch values.

Personally i think having the factor be a variable per ruleset like the multiplier works quite well and could also be beneficial for custom rulesets possibly giving score in the future. (thinking of touhosu, which also has a lot more hit objects)

**Comment by respektive** [comment, marked as answer]
I have done some testing using a single multiplier per ruleset like mentioned in [this comment](https://github.com/ppy/osu/discussions/17125#discussioncomment-2305327).  
The formula would look like this instead then, replacing the fixed multiplier value:
![](https://cdn.discordapp.com/attachments/951246438594134057/951250518288773222/142041528875876352.png)

For the ruleset multiplier i settled for these values:

- osu: 36 (same as before)
- taiko: 22
- catch 28
- mania: 16

In my testing these brought the best results. I tried to keep the values somewhat similar across rulesets so beatmaps that give 50mil score in osu don't give 500mil in mania for example.

Here is a video example of how it looks switching between rulesets on current release:

https://user-images.githubusercontent.com/22381687/157556350-e7da8ce5-ceec-4c61-8084-ef1fe6ac22d0.mp4

And here using my values:

https://user-images.githubusercontent.com/22381687/157556369-b2528134-3229-4ba9-93c2-6cb9efc941ba.mp4

I also tested these values on the total rankings by recalculating the scores in the [database dumps](https://data.ppy.sh/) released by peppy.

osu!catch:
| user\_id  | stable values   | \*36 multiplier | \*28 multiplier |
| --------- | --------------- | --------------- | --------------- |
| 2400918   | 659,015,236,533 | 758,271,459,482 | 591,705,670,282 |
| 1375955   | 591,044,646,959 | 751,465,044,041 | 587,692,658,802 |
| 448547    | 549,722,973,008 | 693,203,299,613 | 541,177,755,648 |
| 2351567   | 494,434,653,531 | 599,519,057,352 | 467,522,439,123 |
| 4568537   | 413,197,369,418 | 513,929,048,534 | 401,085,059,485 |
| 1101600   | 430,873,722,923 | 505,333,155,890 | 394,581,703,007 |
| 538717    | 360,080,011,038 | 460,544,190,702 | 360,077,239,767 |
| 13279756  | 346,574,024,766 | 433,229,888,162 | 338,420,622,289 |
| 5163523   | 283,537,815,421 | 363,730,913,172 | 284,046,088,927 |

osu!taiko:
| user\_id  | stable values  | \*36 multiplier | \*22 multiplier |
| --------- | -------------- | --------------- | --------------- |
| 6503700   | 27,474,827,419 | 769,837,118,105 | 473,516,787,001 |
| 374950    | 24,924,639,747 | 431,667,015,143 | 270,031,207,845 |
| 3165416   | 13,474,369,309 | 418,799,588,861 | 257,933,428,096 |
| 4799788   | 13,827,878,952 | 404,907,849,480 | 249,575,153,254 |
| 1292021   | 11,391,913,911 | 395,281,077,958 | 242,817,805,626 |
| 9912966   | 9,388,675,097  | 381,106,432,324 | 232,919,428,116 |
| 1083444   | 16,646,728,978 | 355,439,654,672 | 220,595,869,174 |
| 3196091   | 9,134,361,726  | 354,251,685,267 | 217,126,934,018 |
| 3802922   | 10,202,758,580 | 345,030,124,913 | 210,868,770,704 |

osu!mania:
| user\_id  | stable values  | \*36 multiplier   | \*16 multiplier   |
| --------- | -------------- | ----------------- | ----------------- |
| 4937439   | 87,619,527,041 | 2,504,437,113,350 | 1,113,083,161,489 |
| 5818319   | 5,006,553,331  | 982,109,694,232   | 436,493,197,436   |
| 7712274   | 24,210,686,337 | 868,444,877,961   | 385,975,501,316   |
| 6914714   | 5,040,591,826  | 812,636,975,552   | 361,171,989,134   |
| 3357640   | 8,866,416,428  | 782,527,882,039   | 347,790,169,795   |
| 6694908   | 5,007,137,494  | 732,363,236,687   | 325,494,771,861   |
| 3338246   | 9,512,018,306  | 718,172,512,611   | 319,187,783,382   |
| 1517607   | 6,935,444,506  | 667,109,019,994   | 296,492,897,775   |
| 11247298  | 6,711,739,923  | 616,182,996,384   | 273,859,109,504   |

I think these numbers look a lot better, especially on mania where the current #1 player would literally gain trillions.

**Reply by smoogipoo** [reply]
Nice work. I think these values look good. PR here: https://github.com/ppy/osu/pull/17195

**Reply by peppy** [reply]
I'm curious about the total scores still being up to a magnitude off - is there a reason not to drop the multiplier further to handle these scenarios? Is total score somehow affected differently from per beatmap scores?

**Reply by respektive** [reply]
My idea was to keep the values for taiko and mania similar to how they are in osu and catch, so it doesn't feel odd when you switch to another ruleset and the scores are completely different values. So the total score values do increase quite a lot, because of this, which is something i was not too sure about as well.

I have asked some other members of the score farmer community, especially ones that are playing the other modes more than i do, and from what i heard, there is no real point to drop the multipliers further to get them similar to scorev1 if it doesn't change much. 

Having a bigger multiplier does not change the order on the rankings and it feels more natural if all rulesets have similar score values. If we were to drop it further to get them closer to the current values then we might as well not use the classic scoring for taiko and mania at all, since right now mania is capped at 1mil anyway and taiko is not that much different either in terms of score values.
